### PR TITLE
feat: `hmac-auth` add signed headers to  calculate signature

### PR DESCRIPTION
**Pull request by nic-chen**
### What this PR does / why we need it:

`hmac-auth` add signed headers to  calculate signature

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/plugins/hmac-auth.lua [hunk 1]
```diff
@@ -52,16 +58,19 @@ local schema = {
...
                     type = "integer",
                     default = 300
-                }
+                },
+                signed_headers = {
+                    type = "array",
+                    items = {
+                        type = "string",
+                        minLength = 1,
+                        maxLength = 20,
```
Comment (gxthrj) [inline apisix/plugins/hmac-auth.lua:67]: The maxLength of 20 is not enough, how about 50?

Comment (nic-chen) [reply apisix/plugins/hmac-auth.lua:67]: ok

File: apisix/plugins/hmac-auth.lua [hunk 2]
```diff
@@ -182,13 +201,25 @@ local function generate_signature(ctx, secret_key, params)
         canonical_query_string = core.table.concat(query_tab, "&")
     end
 
-    local req_body = core.request.get_body()
-    req_body = req_body or ""
+    local canonical_headers = ""
+
+    core.log.info("all headers: ", core.json.encode(core.request.headers(), true))
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:206]: `core.request.headers(ctx)`
we should pass the `ctx` for better performance

File: apisix/plugins/hmac-auth.lua [hunk 3]
```diff
@@ -182,13 +201,25 @@ local function generate_signature(ctx, secret_key, params)
...
+
+    for _, h in pairs(params.signed_headers) do
+        canonical_headers = canonical_headers .. (core.request.header(ctx, h) or "")
+        core.log.info("canonical_headers:", canonical_headers,
+        " header:", core.json.encode(h),
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:211]: bad indentation

File: apisix/plugins/hmac-auth.lua [hunk 4]
```diff
@@ -182,13 +201,25 @@ local function generate_signature(ctx, secret_key, params)
         canonical_query_string = core.table.concat(query_tab, "&")
     end
 
-    local req_body = core.request.get_body()
-    req_body = req_body or ""
+    local canonical_headers = ""
+
+    core.log.info("all headers: ", core.json.encode(core.request.headers(), true))
+
+    for _, h in pairs(params.signed_headers) do
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:208]: I think we should use `ipairs` here

File: apisix/plugins/hmac-auth.lua [hunk 5]
```diff
@@ -182,13 +201,25 @@ local function generate_signature(ctx, secret_key, params)
...
+        canonical_headers = canonical_headers .. (core.request.header(ctx, h) or "")
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:209]: `canonical_headers` can be an array object, then we can use `table.insert` and `table.concat` way, which is better performance

File: apisix/plugins/hmac-auth.lua [hunk 6]
```diff
@@ -264,18 +310,22 @@ local function get_params(ctx)
...
             timestamp = auth_data[5]
+            signed_headers = auth_data[6]
         end
     end
 
     params.access_key = ak
     params.algorithm  = algorithm
     params.signature  = signature
     params.timestamp  = timestamp or 0
+    params.signed_headers = ngx_re.split(signed_headers or "", ";")
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:326]: if the `signed_headers` is `nil`, I think `params.signed_headers` is `nil`.

it is an optional field.

**Comment by membphis** [comment]
@nic-chen you should update the doc, we can confirm the doc first

**Comment by nic-chen** [comment]
@membphis  @gxthrj  Chinese doc has updated. please take a look.

File: doc/zh-cn/plugins/hmac-auth.md [hunk 7]
```diff
@@ -41,6 +41,8 @@
 | `secret_key`| 必须 | 无 | 与 `access_key` 配对使用。|
 | `algorithm` | 可选 | hmac-sha256 | 加密算法。目前支持 `hmac-sha1`, `hmac-sha256` 和 `hmac-sha512`。|
 | `clock_skew`| 可选 | 300 | 签名允许的时间偏移，以秒为单位的计时。比如允许时间偏移 10 秒钟，那么就应设置为 `10`。特别地，`0` 表示不对 `timestamp` 进行检查。|
+| `signed_headers` | 可选 | 无 | 限制加入加密计算的 headers ，指定后客户端请求只能在此范围内指定 headers ，此项为空时将把所有客户端请求指定的 headers 加入加密计算。|
```
Comment (membphis) [inline doc/zh-cn/plugins/hmac-auth.md:44]: add some example

Comment (nic-chen) [reply doc/zh-cn/plugins/hmac-auth.md]: > add some example

updated

File: apisix/plugins/hmac-auth.lua (hunk 5 above)
Comment (nic-chen) [reply apisix/plugins/hmac-auth.lua]: not quite understand here, this is concatenating a string to generate a signature

Comment (membphis) [reply apisix/plugins/hmac-auth.lua]: I helped you fix it. you can take a look when you have time

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@gxthrj  do you have time to take a look at this PR?

**Comment by gxthrj** [comment]
> @gxthrj do you have time to take a look at this PR?

ok

**Review by gxthrj — APPROVED** [review]

**Pull request merged**